src/typeChecker/typeErrorChecker.ts

```diff
@@ -0,0 +1,914 @@
+          throw new TypecheckError(node, 'Function declaration should always have an identifier')
+        }
+        // Only identifiers are used as function params in Source
+        const params = node.params as tsEs.Identifier[]
+        const returnType = getTypeAnnotationType(node.returnType, context, env)
+
+        const types = getParamTypes(params, context, env)
+        // Return type will always be last item in types array
+        types.push(returnType)
+        const fnType = tFunc(...types)
+
+        // Save function type in type env
+        setType(node.id.name, fnType, env)
+        break
+      case 'VariableDeclaration':
+        if (node.kind === 'var') {
+          throw new TypecheckError(node, 'Variable declaration using "var" is not allowed')
+        }
+        if (node.declarations.length !== 1) {
+          throw new TypecheckError(
+            node,
+            'Variable declaration should have one and only one declaration'
+          )
+        }
+        if (node.declarations[0].id.type !== 'Identifier') {
+          throw new TypecheckError(node, 'Variable declaration ID should be an identifier')
+        }
+        const id = node.declarations[0].id as tsEs.Identifier
+        const expectedType = getTypeAnnotationType(id.typeAnnotation, context, env)
+
+        // Save variable type and decl kind in type env
+        setType(id.name, expectedType, env)
+        setDeclKind(id.name, node.kind, env)
+        break
+      case 'TSTypeAliasDeclaration':
+        const type = getTypeAnnotationType(node, context, env)
+        setTypeAlias(node.id.name, type, env)
+        break
+      default:
+        break
+    }
+  })
+}
+
+/**
+ * Typechecks the body of a binary expression, adding any type errors to context if necessary.
+ * Then, returns the type of the binary expression, inferred based on the operator.
+ */
+function typeCheckAndReturnBinaryExpressionType(
+  node: tsEs.BinaryExpression,
+  context: Context,
+  env: TypeEnvironment
+): Type {
+  const leftType = typeCheckAndReturnType(node.left, context, env)
+  const rightType = typeCheckAndReturnType(node.right, context, env)
+  const leftTypeString = formatTypeString(leftType)
+  const rightTypeString = formatTypeString(rightType)
+  const operator = node.operator
+  switch (operator) {
+    case '-':
+    case '*':
+    case '/':
+    case '%':
+      // Return type number
+      checkForTypeMismatch(node, leftType, tNumber, context)
+      checkForTypeMismatch(node, rightType, tNumber, context)
+      return tNumber
+    case '+':
+      // Both sides can only be number, string, or any
+      // However, case where one side is string and other side is number is not allowed
+      if (leftTypeString === 'number' || leftTypeString === 'string') {
+        checkForTypeMismatch(node, rightType, leftType, context)
+        // If string + number, return string; else return left type
```

Comment:
So, is one side being string and the other side being number allowed or not allowed? These comments seem to contradict each other.